lib/handler/cache-revalidation-handler.js

```diff
@@ -0,0 +1,105 @@
+'use strict'
+
+const DecoratorHandler = require('../handler/decorator-handler')
+
+/**
+ * This takes care of revalidation requests we send to the origin. If we get
+ *  a response indicating that what we have is cached (via a HTTP 304), we can
+ *  continue using the cached value. Otherwise, we'll receive the new response
+ *  here, which we then just pass on to the next handler (most likely a
+ *  CacheHandler). Note that this assumes the proper headers were already
+ *  included in the request to tell the origin that we want to revalidate the
+ *  response (i.e. if-modified-since).
+ *
+ * @see https://www.rfc-editor.org/rfc/rfc9111.html#name-validation
+ *
+ * @typedef {import('../../types/dispatcher.d.ts').default.DispatchHandlers} DispatchHandlers
+ * @implements {DispatchHandlers}
+ */
+class CacheRevalidationHandler extends DecoratorHandler {
+  #successful = false
+  /**
+   * @type {() => void}
+   */
+  #successCallback = null
+  /**
+   * @type {import('../../types/dispatcher.d.ts').default.DispatchHandlers}
+   */
+  #handler = null
+
+  /**
+   * @param {() => void} successCallback Function to call if the cached value is valid
+   * @param {import('../../types/dispatcher.d.ts').default.DispatchHandlers} handler
+   */
+  constructor (successCallback, handler) {
+    super(handler)
+
+    this.#successCallback = successCallback
+    this.#handler = handler
+  }
+
+  /**
+   * @see {DispatchHandlers.onHeaders}
+   *
+   * @param {number} statusCode
+   * @param {Buffer[]} rawHeaders
+   * @param {() => void} resume
+   * @param {string} statusMessage
+   * @param {string[] | undefined} headers
+   * @returns {boolean}
+   */
+  onHeaders (
+    statusCode,
+    rawHeaders,
+    resume,
+    statusMessage,
+    headers = undefined
+  ) {
+    // https://www.rfc-editor.org/rfc/rfc9111.html#name-handling-a-validation-respo
+    if (statusCode === 304) {
+      this.#successful = true
+      this.#successCallback()
+      return true
+    }
+
+    return this.#handler.onHeaders(
```

Comment:
what if #handler has no onHeaders attribute?

Reply:
the function would always be there as marks the kick-off for processing the response; shouldn't be possible to find a non `onHeaders` (the whole process will be wrong if not provided). Same to `handler.onError`

Reply:
The type does allow for it and onError to be undefined however https://github.com/nodejs/undici/blob/0ad6007ae74ce2d392c76a90b07e6cda725d3bad/types/dispatcher.d.ts#L226